xarray/core/computation.py

```diff
@@ -926,6 +927,108 @@ def earth_mover_distance(first_samples,
         return apply_array_ufunc(func, *args, dask=dask)
 
 
+def dot(*arrays, **kwargs):
+    """ dot(*arrays, dims=None)
+
+    Generalized dot product for xarray objects. Like np.einsum, but
+    provides a simpler interface based on array dimensions.
+
+    Parameters
+    ----------
+    arrays: DataArray objects
+        Arrays to compute.
+    dims: str or tuple of strings, optional
+        Which dimensions to sum over.
+        If not speciified, then all the common dimensions are summed over.
+
+    Returns
+    -------
+    dot: DataArray
+
+    Examples
+    --------
+
+    >>> da_a = xr.DataArray(np.arange(3 * 4).reshape(3, 4), dims=['a', 'b'])
+    >>> da_b = xr.DataArray(np.arange(3 * 4 * 5).reshape(3, 4, 5),
+    >>>                     dims=['a', 'b', 'c'])
+    >>> da_c = xr.DataArray(np.arange(5 * 6).reshape(5, 6), dims=['c', 'd'])
+    >>>
+    >>> xr.dot(da_a, da_b, dims=['a', 'b']).dims
+    ('c', )
+    >>> xr.dot(da_a, da_b, dims=['a']).dims
+    ('b', 'c')
+    >>> xr.dot(da_a, da_b, da_c, dims=['b', 'c']).dims
+    ('a', 'd')
+    """
+    from .dataarray import DataArray
+
+    dims = kwargs.pop('dims', None)
+    if len(kwargs) > 0:
+        raise TypeError('Invalid keyward arguments {} are given'.format(
+            list(kwargs.keys())))
+
+    if any(not isinstance(arr, DataArray) for arr in arrays):
+        raise TypeError('Only xr.DataArray and xr.Variable are supported.'
+                        'Given {}.'.format([type(arr) for arr in arrays]))
+
+    if len(arrays) == 0:
+        raise TypeError('At least one array should be given.')
+
+    if isinstance(dims, basestring):
+        dims = (dims, )
+    elif isinstance(dims, list):
+        dims = tuple(dims)
```

Comment:
FWIW `dims=tuple(dims)` doesn't create any copies if `dims` is already a tuple, so you could skip the `if isinstance` check